lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -1126,39 +1125,33 @@ newDBLayerWith cacheBehavior ti SqliteContext{runQuery} = do
             NoCache -> pure ()
             CacheLatestCheckpoint -> modifyMVar_ cache (pure . m)
 
-    let writeCache :: W.WalletId -> Maybe (W.Wallet s) -> SqlPersistT IO ()
-        writeCache wid = maybeUpdateCache . \case
-                Nothing -> Map.delete wid
-                Just cp ->
-                    let tip = cp ^. #currentTip . #blockHeight
-                        alter = \case
-                            Just old | tip < old ^. #currentTip .  #blockHeight ->
-                                Just old
-                            _ ->
-                                Just cp
-                    in Map.alter alter wid
-
-    let selectLatestCheckpoint
+        writeCache :: W.WalletId -> Maybe (W.Wallet s) -> SqlPersistT IO ()
+        writeCache wid = maybeUpdateCache . flip Map.alter wid . maybe (const Nothing) alterCache
+
+        alterCache :: W.Wallet s -> (Maybe (W.Wallet s) -> Maybe (W.Wallet s))
+        alterCache cp = \case
+            -- this seems suspicious
+            Just old | getHeight cp < getHeight old -> Just old
```

Comment:
This is mainly there because of the state-machine tests. In the application, we do in practice never have the case where we try to write a checkpoint that is older than the latest checkpoint written because we process them in sequence from genesis to the tip, and we drop checkpoints upon rolling back. 

However, the state-machine tests aren't really respecting any of this and may insert checkpoints in any arbitrary order. That little guard is here to cope with this and enforce the invariant that the cache always contains the _most recent checkpoint_, not in terms of insertion order in the database but in terms of blockheight on the chain. 

I realize that this deserves a proper comment note...